ttnn/cpp/ttnn/operations/ccl/reduce_scatter/host/reduce_scatter_common.cpp

```diff
@@ -0,0 +1,173 @@
+        worker_receiver_semaphore_id_second_core_range;
+
+    for (std::size_t l = 0; l < num_links; l++) {
+        for (std::size_t i = 0; i < workers_per_slice; i++) {
+            auto worker_id = get_global_worker_id(l, i, num_channels_per_link);
+            TT_ASSERT(worker_cores_idx < worker_cores_list.size());
+
+            worker_attributes.push_back(
+                {
+                    l,
+                    i,
+                    i,
+                    is_buffer_in_clockwise_direction_fn(worker_id) ? Direction::CLOCKWISE : Direction::COUNTER_CLOCKWISE,
+                    first_workers_list[worker_cores_idx],
+                    first_send_to_edm_sem_id,
+                    first_read_from_edm_sem_id
+                }
+            );
+            log_trace(tt::LogOp, "Worker {} direction= {}", i, worker_attributes.back().direction == Direction::CLOCKWISE ? "CLOCKWISE" : "COUNTER-CLOCKWISE");
+            worker_cores_idx++;
+        }
+        if (topology_config.is_linear) {
+            auto & second_vec_index = split_grids ? second_worker_cores_idx : worker_cores_idx;
+            for (std::size_t i = 0; i < workers_per_slice; i++) {
+                TT_ASSERT(second_vec_index < second_workers_list.value().size());
+                std::size_t my_logical_index = workers_per_slice + i;
+                std::size_t my_idx = worker_attributes.size();
+                worker_attributes.push_back(
+                    {
+                        l,
+                        my_logical_index,
+                        i,
+                        is_buffer_in_clockwise_direction_fn(my_logical_index) ?
+                            Direction::CLOCKWISE : Direction::COUNTER_CLOCKWISE,
+                        second_workers_list.value()[second_vec_index],
+                        second_send_to_edm_sem_id,
+                        second_read_from_edm_sem_id
+                    }
+                );
+                log_trace(tt::LogOp, "Worker {} direction= {}", my_logical_index, worker_attributes.back().direction == Direction::CLOCKWISE ? "CLOCKWISE" : "COUNTER-CLOCKWISE");
+                std::size_t associated_idx = my_idx - workers_per_slice;
+                worker_attributes[my_idx].associated_worker_index = associated_idx;
+                worker_attributes[my_idx].associated_worker_core_logical = worker_attributes[associated_idx].location_logical;
+                worker_attributes[associated_idx].associated_worker_index = my_idx;
+                worker_attributes[associated_idx].associated_worker_core_logical = worker_attributes[my_idx].location_logical;
+                second_vec_index++;
+            }
+        }
+    }
+
+    // Validate the worker attributes
+    for (const auto &wa : worker_attributes) {
+        TT_ASSERT(wa.send_to_edm_semaphore_id.has_value() || wa.receive_from_edm_semaphore_id.has_value(), "Internal error. Incorrectly setup worker attributes for reduce scatter");
```

Comment:
Why is it an assert vs fatal here?

Reply:
Yea this is one of those things with assert vs fatal. Here's how I see it - let me know if you disagree.

If it's an external user facing API or some variable directly derived from some user provided information --> Always fatal if we know the provided input is incorrect/will provide issues with the program.

If it's an internal logic bug we are trying to catch --> Use TT_ASSERT as we should be catching these during testing... Unless these are disabled during our test builds too  . In this case I am trying to catch the internal logic bug category of error.